src/d3d9/d3d9_shader_validator.h

```diff
@@ -37,32 +71,203 @@ namespace dxvk {
+      }*/
+
+      // a maximum of 10 inputs are supported with PS 3.0 (validation required by The Void)
+      if (m_isPixelShader && m_majorVersion == 3) {
+        switch (instContext.instruction.opcode) {
+          case DxsoOpcode::Comment:
+          case DxsoOpcode::Def:
+          case DxsoOpcode::DefB:
+          case DxsoOpcode::DefI:
+            break;
+
+          default:
+            for (uint32_t instNum = 1; pdwInst[instNum] & 0x80000000; instNum++) {
+              DWORD regType  = ((pdwInst[instNum] & D3DSP_REGTYPE_MASK)  >> D3DSP_REGTYPE_SHIFT)
+                             | ((pdwInst[instNum] & D3DSP_REGTYPE_MASK2) >> D3DSP_REGTYPE_SHIFT2);
+              DWORD regIndex = pdwInst[instNum] & D3DSP_REGNUM_MASK;
+
+              if (unlikely(regType == D3DSPR_INPUT && regIndex >= 10)) {
+                return ErrorCallback(pFile, Line, 0x2,
+                  instContext.instruction.opcode == DxsoOpcode::Dcl ?
+                    D3D9ShaderValidatorMessage::BadInputRegisterDeclaration :
+                    D3D9ShaderValidatorMessage::BadInputRegister,
+                  "IDirect3DShaderValidator9::Instruction: Invalid number of PS input registers specified. Aborting validation.");
+              }
+            }
+        }
+      }
 
       return D3D_OK;
     }
 
 
     HRESULT STDMETHODCALLTYPE End() {
-      Logger::debug("D3D9ShaderValidator::End: Stub");
+      if (unlikely(m_state == D3D9ShaderValidatorState::Error)) {
+        return E_FAIL;
+      } else if (unlikely(m_state == D3D9ShaderValidatorState::Begin)) {
+        return ErrorCallback(nullptr, 0, 0,
+          D3D9ShaderValidatorMessage::EndOutOfOrder,
+          "IDirect3DShaderValidator9::End called out of order. Call to ::Begin, followed by calls to ::Instruction must occur first.");
+      } else if (unlikely(m_state != D3D9ShaderValidatorState::EndOfShader)) {
+        return ErrorCallback(nullptr, 0, 0,
+          D3D9ShaderValidatorMessage::MissingEndToken,
+          "IDirect3DShaderValidator9::End: Shader missing end token.");
+      }
+
+      m_state    = D3D9ShaderValidatorState::Begin;
+      m_isPixelShader = false;
+      m_majorVersion  = 0;
+      m_minorVersion  = 0;
+      m_callback = nullptr;
+      m_userData = nullptr;
+      m_ctx      = nullptr;
 
       return D3D_OK;
     }
 
+  private:
+
+    HRESULT ValidateHeader(const char* pFile, UINT Line, const DWORD* pdwInst, DWORD cdw) {
+      if (unlikely(cdw != 1)) {
+        return ErrorCallback(pFile, Line, 0x6,
+          D3D9ShaderValidatorMessage::BadVersionTokenLength,
+          "IDirect3DShaderValidator9::Instruction: Bad version token. DWORD count > 1 given. Expected DWORD count to be 1 for version token.");
+      }
+
+      DxsoReader reader = { reinterpret_cast<const char*>(pdwInst) };
+      uint32_t headerToken = reader.readu32();
+      uint32_t shaderType  = headerToken & 0xffff0000;
```

Comment:
I'd prefer it if we could reuse `DxsoHeader` & `DxsoProgramInfo` here unless there's a good reason for why that doesn't work.

You need `DxsoProgramInfo` later anyway so you might as well let the constructor of `DxsoHeader` read out the program type.

Reply:
No other reason than I didn't quite know they existed (tried to reuse our own Dxso stuff as much as possible, but it's a brave new world for me). Will give it a go.

Reply:
As discussed on Discord: `DxsoHeader` throws an exception when it encounters invalid values which ofc isn't great for the validator. So we'll keep this as it is atm.

Reply:
`DxsoHeader` will thrown an exception for invalid versions, so I'd rather not use it here. As far as `DxsoProgramInfo` goes, we are creating one futher down (and saving it under m_ctx).